src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,418 @@
+    // This is an error
+    Kokkos::abort("Trying to send from a receiver");
+  }
+}
+
+template <class T>
+void CommBuffer<T>::SendNull() noexcept {
+  PARTHENON_DEBUG_REQUIRE(*state_ == BufferState::stale,
+                          "Trying to send_null from buffer that hasn't been staled.");
+  *state_ = BufferState::sending_null;
+  if (*comm_type_ == BuffCommType::sender) {
+// Make sure that this request isn't still out,
+// this could be blocking
+#ifdef MPI_PARALLEL
+    PARTHENON_MPI_CHECK(MPI_Wait(my_request_.get(), MPI_STATUS_IGNORE));
+    PARTHENON_MPI_CHECK(MPI_Isend(&null_buf_, 0, MPITypeMap<buf_base_t>::type(),
+                                  recv_rank_, tag_, comm_, my_request_.get()));
+#endif
+  }
+  if (*comm_type_ == BuffCommType::receiver) {
+    // This is an error
+    Kokkos::abort("Trying to send from a receiver");
+  }
+}
+
+template <class T>
+bool CommBuffer<T>::IsAvailableForWrite() {
+  if (*comm_type_ == BuffCommType::sender) {
+#ifdef MPI_PARALLEL
+    // We do not check stale status here since the receiving end should be the one
+    // setting the buffer to stale, all we care about for a pure sender is wether
+    // or not its last send message has been completed
+    if (*state_ == BufferState::stale) return true;
+    if (*my_request_ == MPI_REQUEST_NULL) return true;
+    int flag, test;
+    PARTHENON_MPI_CHECK(MPI_Iprobe(MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &test,
+                                   MPI_STATUS_IGNORE));
+    PARTHENON_MPI_CHECK(MPI_Test(my_request_.get(), &flag, MPI_STATUS_IGNORE));
+    if (flag) *state_ = BufferState::stale;
+    return flag;
+#else
+    PARTHENON_FAIL("Should not have a sending buffer when MPI is not enabled.");
+#endif
+  } else if (*comm_type_ == BuffCommType::both) {
+    if (*state_ == BufferState::stale) return true;
+    return false;
+  } else {
+    PARTHENON_FAIL("Receiving buffer is never available for write.");
+  }
+}
+
+template <class T>
+void CommBuffer<T>::TryStartReceive() noexcept {
+#ifdef MPI_PARALLEL
+  if (*comm_type_ == BuffCommType::receiver && !*started_irecv_) {
+    PARTHENON_REQUIRE(
+        *my_request_ == MPI_REQUEST_NULL,
+        "Cannot have another pending request in a buffer that is starting to receive.");
+    if (!IsActive())
+      Allocate(); // For early start of Irecv, always need storage space even if not used
+    PARTHENON_MPI_CHECK(MPI_Irecv(buf_.data(), buf_.size(),
+                                  MPITypeMap<buf_base_t>::type(), send_rank_, tag_, comm_,
+                                  my_request_.get()));
+    *started_irecv_ = true;
+  } else if (*comm_type_ == BuffCommType::sparse_receiver && !*started_irecv_) {
+    int test;
+    MPI_Status status;
+    // This is the extra MPI call that impacts performance mentioned in Athena++
+    // PARTHENON_MPI_CHECK(MPI_Iprobe(MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &test,
+    //                               MPI_STATUS_IGNORE));
```

Comment:
Maybe reintroduce the full original comment?

Reply:
I put it back in